### PR TITLE
Wrap provider errors with HTTP status code

### PR DESCRIPTION
**Pull request by biglittlebigben**
## Summary
- Add `ErrorWithStatusCode` type so callers can extract HTTP status codes via `errors.As` regardless of the storage backend.
- Wrap returned errors from S3, GCP, Azure, and AliOSS storage operations to expose status codes uniformly.
- Add unit tests covering nil, native, wrapped, and pass-through cases for each backend.

## Motivation
Enables better metrics on upload/download failures by surfacing the HTTP status code in a backend-agnostic way — in particular, distinguishing user errors (4xx: bad credentials, missing object, access denied) from infrastructure issues (5xx, network errors that pass through unwrapped).

## Test plan
- [ ] `go test ./...`

### CODE REVIEW

**Review by paulwe — APPROVED** [review]

**Review by boks1971 — APPROVED** [review]

File: error.go [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+type ErrorWithStatusCode struct {
+	Err        error
+	StatusCode int
+}
+
+func (e *ErrorWithStatusCode) Error() string {
+	return fmt.Sprintf("Err: %s, Status Code: %d", e.Err, e.StatusCode)
+}
+
+func (e *ErrorWithStatusCode) UnWrap() error {
```
Comment (milos-lk) [inline error.go]: should this be `Unwrap` instead?

Comment (milos-lk) [reply error.go]: maybe we could add assertion on the wrapped error in tests as well

File: error.go [hunk 2]
```diff
@@ -0,0 +1,18 @@
+package storage
```
Comment (milos-lk) [inline error.go:15]: license header missing

File: error.go [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+var _ error = (*ErrorWithStatusCode)(nil)
+
+type ErrorWithStatusCode struct {
+	Err        error
+	StatusCode int
+}
+
+func (e *ErrorWithStatusCode) Error() string {
+	return fmt.Sprintf("Err: %s, Status Code: %d", e.Err, e.StatusCode)
```
Comment (milos-lk) [inline error.go]: nit: error stings are usually not capitalized

**Review by milos-lk — APPROVED** [review]
